### подключение

In [56]:
import pandas as pd
import psycopg2

DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

### отключение

In [112]:
connection.close()

# задание на знакомство с данными 

In [111]:
#3.1-4
query = f"""
SELECT COUNT(DISTINCT id) FROM vacancies;
--SELECT COUNT(DISTINCT id) FROM employers;
--SELECT COUNT(DISTINCT id) FROM areas;
--SELECT COUNT(DISTINCT id) FROM industries;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\2392454814.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,49197


Вывод: база данных содержит 49 197 вакансий, 23 501 работодателя, 1 362 города и 294 сферы деятельности. Это говорит о том, что данные достаточно объёмные и разнообразные: по ним можно анализировать не только отдельные профессии, но и структуру рынка труда в целом, сравнивать регионы, отрасли и типы работодателей.

In [ ]:
#4.1
query = f"""
SELECT a.name AS area, COUNT(v.id) AS cnt
FROM vacancies v
JOIN areas a ON v.area_id = a.id
GROUP BY a.name
ORDER BY cnt DESC;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_14636\1102325622.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,area,cnt
0,Москва,5333
1,Санкт-Петербург,2851
2,Минск,2112
3,Новосибирск,2006
4,Алматы,1892
...,...,...
764,Тарко-Сале,1
765,Новоаннинский,1
766,Бирск,1
767,Сасово,1


In [15]:
#4.2
query = f"""
SELECT COUNT(*)
FROM vacancies
WHERE salary_from IS NOT NULL
   OR salary_to IS NOT NULL;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_14636\1469084257.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,24073


In [16]:
#4.3
query = f"""
SELECT ROUND(AVG(salary_from)) AS avg_salary_from,
       ROUND(AVG(salary_to)) AS avg_salary_to
FROM vacancies;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_14636\3552471519.py:7: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,avg_salary_from,avg_salary_to
0,71065.0,110537.0


In [17]:
#4.4
query = f"""
SELECT schedule, employment, COUNT(*) AS cnt
FROM vacancies
GROUP BY schedule, employment
ORDER BY cnt DESC;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_14636\131354586.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,schedule,employment,cnt
0,Полный день,Полная занятость,35367
1,Удаленная работа,Полная занятость,7802
2,Гибкий график,Полная занятость,1593
3,Удаленная работа,Частичная занятость,1312
4,Сменный график,Полная занятость,940
5,Полный день,Стажировка,569
6,Вахтовый метод,Полная занятость,367
7,Полный день,Частичная занятость,347
8,Гибкий график,Частичная занятость,312
9,Полный день,Проектная работа,141


In [18]:
#4.5
query = f"""
SELECT experience
FROM vacancies
GROUP BY experience
ORDER BY COUNT(*) ASC;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_14636\2547193945.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,experience
0,Более 6 лет
1,Нет опыта
2,От 3 до 6 лет
3,От 1 года до 3 лет


В базе видно, что рынок труда в первую очередь сосредоточен в крупных городах и ориентирован на специалистов с уже имеющимся опытом. Зарплаты раскрываются не всегда, но там, где они указаны, диапазон довольно широкий, что отражает сильную зависимость от уровня кандидата и условий работы. В целом база выглядит как хороший срез современного IT- и аналитического рынка: здесь много вакансий в крупных центрах, заметна роль удалёнки, а основной спрос приходится на кандидатов среднего уровня.

In [19]:
#5.1
query = f"""
SELECT e.name, COUNT(v.id) AS cnt
FROM employers e
JOIN vacancies v ON v.employer_id = e.id
GROUP BY e.name
ORDER BY cnt DESC
LIMIT 5;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_14636\116396589.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,name,cnt
0,Яндекс,1933
1,Ростелеком,491
2,Тинькофф,444
3,СБЕР,428
4,Газпром нефть,331


In [4]:
#5.2
query = f"""
SELECT a.name,
       COUNT(e.id) AS employers_cnt
FROM areas a
LEFT JOIN vacancies v ON v.area_id = a.id
LEFT JOIN employers e ON e.area = a.id
WHERE v.id IS NULL
GROUP BY a.id, a.name
ORDER BY employers_cnt DESC
LIMIT 1;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\596640742.py:13: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,name,employers_cnt
0,Россия,410


In [2]:
#5.3
query = f"""
SELECT MAX(cnt)
FROM (
    SELECT employer_id, COUNT(DISTINCT area_id) AS cnt
    FROM vacancies
    GROUP BY employer_id
) t;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_25660\3008025498.py:10: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,max
0,181


In [3]:
#5.4
query = f"""
SELECT COUNT(*)
FROM employers e
LEFT JOIN employers_industries ei ON e.id = ei.employer_id
WHERE ei.industry_id IS NULL;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_25660\3118689441.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,8419


In [4]:
#5.5
query = f"""
SELECT e.name
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
GROUP BY e.id, e.name
HAVING COUNT(DISTINCT ei.industry_id) = 4
ORDER BY e.name
LIMIT 1 OFFSET 2;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_25660\1261664120.py:11: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,name
0,2ГИС


In [5]:
#5.6
query = f"""
SELECT COUNT(DISTINCT e.id)
FROM employers e
JOIN employers_industries ei ON e.id = ei.employer_id
JOIN industries i ON ei.industry_id = i.id
WHERE i.name = 'Разработка программного обеспечения';
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_25660\1863898648.py:9: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,3553


In [19]:
#5.7
query = f"""
WITH y AS (
    SELECT a.name AS area, COUNT(*) AS cnt
    FROM vacancies v
    JOIN employers e ON v.employer_id = e.id
    JOIN areas a ON v.area_id = a.id
    WHERE e.name = 'Яндекс'
      AND a.name IN (
          'Москва', 'Санкт-Петербург', 'Новосибирск', 'Екатеринбург',
          'Казань', 'Красноярск', 'Нижний Новгород', 'Челябинск',
          'Уфа', 'Краснодар', 'Самара', 'Ростов-на-Дону',
          'Омск', 'Воронеж', 'Пермь', 'Волгоград'
      )
    GROUP BY a.name
)
SELECT *
FROM y

UNION ALL

SELECT 'Total' AS area, SUM(cnt) AS cnt
FROM y;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_25660\4017315611.py:25: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,area,cnt
0,Волгоград,24.0
1,Воронеж,32.0
2,Екатеринбург,39.0
3,Казань,25.0
4,Краснодар,30.0
5,Красноярск,23.0
6,Москва,54.0
7,Нижний Новгород,36.0
8,Новосибирск,35.0
9,Омск,21.0


База данных хорошо показывает типичную структуру российского IT-рынка: основные работодатели — это крупные компании и экосистемы, а не малый бизнес. Лидирует Яндекс, заметно от него отстают Ростелеком, Тинькофф, СБЕР и Газпром нефть, что говорит о высокой концентрации вакансий у нескольких крупных игроков.

При этом рынок вакансий распределён неравномерно и по географии, и по отраслям. У части компаний вакансии есть сразу в большом числе городов, а у некоторых работодателей присутствует широкая региональная сеть, что особенно заметно по максимальному числу городов у одного работодателя. Это подтверждает, что большие компании ищут сотрудников не только в столицах, но и в регионах.

Отдельно видно, что многие работодатели вообще не привязаны к одной конкретной отрасли: у значительной части компаний сфера деятельности в базе не заполнена. Там, где отрасль указана, заметную роль играет разработка программного обеспечения, то есть ядро вакансий связано именно с IT.

По Яндексу видно, что он активно нанимает в разных крупных городах, но основная часть вакансий всё же сосредоточена в Москве и Санкт-Петербурге. Это соответствует общей картине рынка: столицы остаются главными центрами притяжения, а регионы скорее дополняют их, чем конкурируют с ними по объёму.

In [6]:
#6.1
query = f"""
SELECT COUNT(*)
FROM vacancies
WHERE name ILIKE '%data%'
   OR name ILIKE '%данн%';
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\894186872.py:8: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,1771


In [66]:
#6.2
query = f"""
SELECT COUNT(*)
FROM vacancies
WHERE (
    lower(name) LIKE '%data scientist%'
    OR lower(name) LIKE '%data science%'
    OR lower(name) LIKE '%исследователь данных%'
    OR (lower(name) LIKE '%ml%' AND lower(name) NOT LIKE '%html%')
    OR lower(name) LIKE '%machine learning%'
    OR lower(name) LIKE '%машинн%обучен%'
)
AND  (
    lower(name) LIKE '%junior%'
    OR experience = 'Нет опыта'
    OR employment = 'Стажировка'
);
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\3483045658.py:19: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,51


In [68]:
#6.3
query = f"""
SELECT COUNT(*)
FROM vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND key_skills IS NOT NULL
AND key_skills <> ''
AND (
    key_skills ILIKE '%SQL%'
    OR key_skills ILIKE '%postgres%'
);
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\3778938533.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,201


In [71]:
#6.4
query = f"""
SELECT COUNT(*)
FROM vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND key_skills IS NOT NULL
AND key_skills <> ''
AND key_skills ILIKE '%Python%';
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\2136763544.py:17: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,count
0,351


In [107]:
#6.5
query = """
SELECT ROUND(AVG(cardinality(string_to_array(key_skills, E'\\t'))), 2)
FROM vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND key_skills IS NOT NULL
AND key_skills <> ''
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\3496040182.py:16: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,round
0,6.41


In [108]:
#6.6
query = """
SELECT experience,
       ROUND(AVG(
           (COALESCE(salary_from, salary_to) + COALESCE(salary_to, salary_from)) / 2.0
       )) AS avg_salary
FROM vacancies
WHERE (
    name ILIKE '%data scientist%'
    OR name ILIKE '%data science%'
    OR name ILIKE '%исследователь данных%'
    OR (name LIKE '%ML%' AND name NOT ILIKE '%HTML%')
    OR name ILIKE '%machine learning%'
    OR name ILIKE '%машинн%обучен%'
)
AND (salary_from IS NOT NULL OR salary_to IS NOT NULL)
GROUP BY experience
ORDER BY experience;
"""
df = pd.read_sql_query(query, connection)
df

C:\Users\aleks\AppData\Local\Temp\ipykernel_672\1119206946.py:20: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query, connection)


,experience,avg_salary
0,Нет опыта,74643.0
1,От 1 года до 3 лет,139675.0
2,От 3 до 6 лет,243115.0


Вакансий, связанных с data science и смежными направлениями, в базе немного по сравнению со всем рынком, но внутри этого сегмента хорошо видно, что спрос концентрируется вокруг уже сложившегося профессионального ядра. Больше всего ценятся кандидаты с опытом от 1 года до 6 лет, а для совсем начинающих позиций рынок заметно уже: вакансий для junior меньше, чем для специалистов с практическим опытом.

По навыкам видно, что работодатели ждут не один узкий инструмент, а полноценный стек. В среднем в DS-вакансиях указывают 6.41 ключевого навыка, а Python встречается намного чаще, чем SQL/Postgres. Это говорит о том, что Python стал почти обязательной базой, тогда как SQL и базы данных — скорее важное дополнение, особенно для аналитических и продуктовых ролей.

Зарплаты растут очень заметно по мере увеличения опыта. Для кандидатов без опыта средняя зарплата около 74.6 тыс., для опыта от 1 до 3 лет — уже 139.7 тыс., а для уровня от 3 до 6 лет — 243.1 тыс. Такой разрыв показывает, что рынок особенно ценит не просто знание инструментов, а способность самостоятельно решать задачи, доводить модели до внедрения и работать с бизнес-кейсами.

Если смотреть на рынок в целом, то он выглядит зрелым и довольно конкурентным: компании чаще ищут не новичков, а специалистов среднего уровня, которые уже умеют работать с данными в реальных проектах. Это объясняет и число навыков в вакансиях, и рост зарплат вместе с опытом, и более узкое окно для входа в профессию.

# Общий вывод 

На рынке DS-вакансий заметна комбинация трёх тенденций: высокий спрос на универсальные навыки, сильная привязка зарплаты к опыту, и концентрация лучших возможностей в крупных городах и удалённых продуктовых командах. В ближайшей перспективе можно ожидать, что спрос будет и дальше смещаться в сторону специалистов, которые умеют не только строить модели, но и доводить их до бизнеса — то есть сочетать Python, SQL, ML и прикладную аналитику.